### PR TITLE
Add 'using' misplacement fixer to code-cleanup

### LINKED ISSUE #44044: Code Cleanup Fixer - 'Using' placement normalization
Add a Code Cleanup fixer to apply `csharp_using_directive_placement` from the editor config file or the 'using' preferences of VS code style configuration.

Right now this rule seem to be applied when VS adds a new using, but it doesn't update the existing ones.
Which means that people with different configuration will generate inconsistent using placements.

A fixer would allow to force that.

### PR DESCRIPTION
**Pull request by weitzhandler**
Fixes #44044

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs [hunk 1]
```diff
@@ -252,6 +246,41 @@ private void Method()
             return AssertCodeCleanupResult(expected, code);
         }
 
+        [Fact]
+        [Trait(Traits.Feature, Traits.Features.CodeCleanup)]
+        public Task FixMisplacedUsingDirectives()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:251]: can we have another test in the oppsite direction.

Comment (weitzhandler) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:251]: what do you mean, outside to inside?

Comment (weitzhandler) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:251]: What do you mean, outside to inside?
I'll do that but it'll require expanding the base test method with another option for that.
Plus adding another option to [*GenerationOptions.cs*](https://github.com/dotnet/roslyn/blob/e16d3906a6d6a560da72144a942dbac7c8dc2e11/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Editing/GenerationOptions.cs).

Comment (weitzhandler) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:251]: Almost all of the tests in that class are one-directional, except for these two: `systemUsingsFirst`, and `separateUsingGroups`.

The problem is that `AddImportPlacement` is internal, and can't be used as a parameter on this test method.

https://github.com/dotnet/roslyn/blob/91c6b154373d0835c69c889ecd27fa6e6cf29dc2/src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs#L255

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:251]: you could always use a boolean on the method level and then translate that to the internal enum value inside the core method.

Comment (weitzhandler) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:251]: There are 8 tests in that file, where only these two are defined in *GenerationOptions.cs*. That file has 55 references. Do you really want me to change that file?

Comment (weitzhandler) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:251]: Even changing the default value inline to `Inside` doesn't do the inverse.

https://github.com/dotnet/roslyn/blob/e16d3906a6d6a560da72144a942dbac7c8dc2e11/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/CodeStyle/CSharpCodeStyleOptions.cs#L265-L276

Comment (weitzhandler) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs]: OK after playing with it for a while, seems that the `Inside` option, doesn't actively move namespaces inside. It's just that it doesn't move out the nested namespaces.

Updated test accordingly.

Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs]: > and can't be used as a parameter on this test method.

In cases like this, one option is to make the method `private protected` instead.

File: src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs [hunk 2]
```diff
@@ -252,16 +249,99 @@ private void Method()
...
+    }
+}
+";
+
+            return AssertCodeCleanupResult(expected, code);
+        }
+
+        [Fact]
+        [Trait(Traits.Feature, Traits.Features.CodeCleanup)]
+        public Task PlaceUsingDirectivesInsideNamespace()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs:289]: 💡 This should include additional tests:

1. Using directives start outside the namespace, with `placeUsingsOutsideNamespace: true`
1. Using directives start outside the namespace, with `placeUsingsOutsideNamespace: false`
1. A pair of tests that start with usings _both_ inside and outside the namespace (one file has multiple using directives), with each of the placement options

File: src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs]: > ... seems that the Inside option, doesn't actively move namespaces inside ...

The test `WhenInsidePreferred_UsingsInCompilationUnit_UsingsMovedAndSystemPlacedFirstIgnored` suggests that the setting is capable of moving the using directives. We should figure out why it is not applying here.

/cc @JoeRobich

File: src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs (hunk 2 above)
Comment (weitzhandler) [reply src/EditorFeatures/CSharpTest/Formatting/CodeCleanupTests.cs]: Done.

**Comment by sharwell** [comment]
Thanks @weitzhandler!

**Pull request merged**